packages/php-wasm/web/src/lib/web-php-endpoint.ts

```diff
@@ -171,4 +173,36 @@ export class WebPHPEndpoint implements IsomorphicLocalPHP {
 	onMessage(listener: MessageListener): void {
 		_private.get(this)!.php.onMessage(listener);
 	}
+
+	/** @inheritDoc @php-wasm/web!WebPHP.addEventListener */
+	addEventListener(
+		eventType: PHPEvent['type'],
+		listener: PHPEventListener
+	): void {
+		_private.get(this)!.php.addEventListener(eventType, listener);
+	}
+
+	/** @inheritDoc @php-wasm/web!WebPHP.removeEventListener */
+	removeEventListener(
+		eventType: PHPEvent['type'],
+		listener: PHPEventListener
+	): void {
+		_private.get(this)!.php.removeEventListener(eventType, listener);
+	}
+
+	atomic(callbackStr: string, args: any[] = []): void {
+		try {
+			// Turn callbackStr into a function
+			const callbackFn = eval(`(${callbackStr})`);
+			if (typeof callbackFn === 'function') {
+				callbackFn(_private.get(this)!.php, ...args);
+			} else {
+				console.error(
+					'Provided string does not evaluate to a function'
+				);
+			}
+		} catch (error) {
+			console.error('Error evaluating the callback string:', error);
+		}
+	}
```

Comment:
@dmsnell I'm hesitant about this change and I wonder if you have any thoughts.

On one hand, it feels wrong in my gut. It opens up a way to execute code in context of a web worker on playground.wordpress.net.

On the other hand, you can already execute custom JS code by simply crafting a WordPress plugin and tampering with workers. So maybe exposing a direct way to `eval()` things is not so bad?

Still, I wonder if you might have any alternative solutions to offer. I initially introduced this function to be able to do this atomically:

1. Disable journalling
2. Apply a FS change from a remote peer
3. Re-enable journalling

I couldn't do that using message passing from the Playground client as that would be async and we would potentially lose some local journal entries.

Reply:
the name is confusing to me. as for security I don't know. the mere act of having a JS-PHP bridge is what opens up connection points, or possibly of running JS code on the Playground's rendered pages?

> I couldn't do that using message passing from the Playground client as that would be async and we would potentially lose some local journal entries.

would be helpful to hear more about this. my first thought is that some async processes have synchronous counterparts. for instance, the code running inside a `Promise` constructor runs synchronously.

Reply:
do we need to take this approach at all? here we're taking two systems - one for intercepting file operations and one for syncing - both of which are mostly based on announcing events, and trying to fit them together.

I wonder instead if the listener part of the sync receive side would do better _inside_ the FS wrapper.

for example, in `fs-journal` we wrap all the functions in Emscripten's filesystem, and save `php[__private__dont__use].FS[name]`. when we receive sync events couldn't we also call the unbound version of these functions, which wouldn't require that we disable journaling?

I'm assuming a similar thing can happen with SQLite. we'd still try and find a way to stop-the-world when applying received updates, but we could get away with that probably by pushing updates into a queue and having the FS or DB drivers avoid performing any new operations until the queues have been purged, while finishing ones they are already working on.

this may not make sense, and I may be misunderstanding a big flow of data in this PR 🤷‍♂️

Reply:
Oh I like that a lot! I just explored it in https://github.com/WordPress/wordpress-playground/pull/727/commits/caf0553ebb8b9f510ed183ffef610b239a6835d5